src/Type/Php/ArrayFilterFunctionReturnTypeReturnTypeExtension.php

```diff
@@ -105,12 +107,22 @@ public function removeFalsey(Type $type): Type
 				} elseif ($isFalsey->maybe()) {
 					$values[$offset] = TypeCombinator::remove($values[$offset], $falseyTypes);
 					$generalize = true;
+				} else {
+					$notEmpty = true;
 				}
 			}
 
 			$filteredArray = new ConstantArrayType(array_values($keys), array_values($values));
 
-			return $generalize ? $filteredArray->generalize() : $filteredArray;
+			if (! $generalize) {
+				return $filteredArray;
+			}
+
+			if (! $notEmpty) {
+				return $filteredArray->generalize();
+			}
+
+			return TypeCombinator::intersect($filteredArray->generalize(), new NonEmptyArrayType());
```

Comment:
This isn't what I suggested. Keep the test but try to fix `ConstantArrayType::generalize()` implementation instead. It just needs to do `TypeCombinator::intersect(new ArrayType(...), new NonEmptyArrayType());` instead on the last line.

Reply:
When looking more into it I realized that your suggestion is actually incorrect. I added a test which would break in your case:

```
function dummy4(?DateTimeInterface $date): void
{
	assertType('DateTimeInterface|false', min(array_filter([$date])));
}
```

Reply:
My suggestion is necessary, non empty constant array needs to be generalized into a non empty general array. But it might not be sufficient to fix this bug and something else needs to be done in the extension too.

Reply:
Alright. Check this [commit](https://github.com/phpstan/phpstan-src/pull/315/commits/fc004825ea197924d31d6cad2421b79f6b4a43c2) please if it is what you want.

It will most likely break the `ArrayFilterFunctionReturnTypeReturnTypeExtension` again so I'll look into it. Just wanna know if the ConstantArrayType fix is correct.

Reply:
It broke a bunch of other tests though...